### ISSUE TITLE
Enable (access modifier) internal ICollectionFixture classes

### ISSUE THREAD

**Issue by johncrim** [xunit/xunit#833]
Based on some testing, it appears that xunit does not support using a class that implements `ICollectionFixture<FixtureType>` from another assembly. This seems reasonable, because test collections are confined to a single assembly; and supporting discovery in any available assembly will almost surely have an undesirable perf impact.

Given this limitation, there will be cases where the collection fixture class is duplicated in multiple test projects - same signature, possibly same name, different namespace. In order to avoid conflicts, it would be nice if internal collection fixture classes could be used (they cannot currently be used). This seems consistent with the definition of a test collection to always be within an assembly. Eg:

```
[CollectionDefinition("MyTestCollection")]
internal class MyTests() {
    ....
```

Alternatively, it would be nice to link to the collection definition type from the test class (instead of using string matching), eg:

```
[CollectionDefinition(typeof(DatabaseIntegrationTestCollection))]
public class MyTests() {
    ....
```

which would presumably allow using test collection fixture definitions from other assemblies.

**Comment by csoltenborn** [comment]
I ran into the exact same problem, but wasn't aware that test collections are confined to single assemblies (at least [this documentation](https://xunit.github.io/docs/shared-context.html#collection-fixture) does not seem to mention that restriction).

I'm not sure whether I agree with this being reasonable, but I am a rather new xUnit user. However, this should
* at least be documented at the above place
* produce a better error message than the current one (which currently appears to be "The following constructor parameters did not have matching fixture data") - it took me quite a while to figure out the actual cause of the problem

But I still would prefer to be able to share classes implementing `ICollectionFixture<FixtureType>` accross assemblies, I must say. My use case is as follows: I have tests running against a database, and that database is copied to the machine executing the tests to improve test isolation. This task is done by the collection fixture and works great in the sense that the database needs to be copied only once for all tests depending on that database. It would be rather nice to have that behavior also across different assemblies, plus having that complete setup at one place and just being able to use it from every test assembly!

**Comment by bradwilson** [comment]
I've updated the documentation.

**Comment by Tasteful** [comment]
Is there any plan to support CollectionDefinition in another assembly?

My case is that I want to deliver a base test class (for the product i working on) that other developers can inherit the base class in their test to start the application correctly.

In my base-class-project I have the following classes
```
    [CollectionDefinition("Application collection")]
    public class ApplicationCollection : ICollectionFixture<ApplicationFixture>
    {
        // This class has no code, and is never created. Its purpose is simply
        // to be the place to apply [CollectionDefinition] and all the
        // ICollectionFixture<> interfaces.
    }

    public class ApplicationFixture 
    {
        public ApplicationFixture()
        {
            // Code to start the application
        }
    }

    [Collection("Application collection")]
    public abstract class ApplicationTestBase
    {
    }
```
Test in other project should then inherit `ApplicationTestBase`.

In this case the `CollectionDefinition` and `Collection` is in the same assembly but it's not the same assembly that contains the tests. 

To not need to scan all assemblies, should it be possible to do either:

1. add extra property on the `CollectionAttribute` with the real type of the collection definition 
2. find the collection definition in the same assembly as the class the implement the `CollectionAttribute`

**Comment by bradwilson** [comment]
Closed for age.
